src/custom/hooks/useTokenLazy.ts

```diff
@@ -0,0 +1,115 @@
+import { useCallback } from 'react'
+
+import { Token } from '@uniswap/sdk-core'
+import { Contract } from '@ethersproject/contracts'
+
+import { useActiveWeb3React } from 'hooks/web3'
+import { getBytes32TokenContract, getTokenContract } from 'hooks/useContract'
+import { parseStringOrBytes32 } from 'hooks/Tokens'
+import { useAddUserToken } from 'state/user/hooks'
+import { Erc20 } from 'abis/types'
+import { retry } from 'utils/retry'
+
+const contractsCache: Record<string, Erc20> = {}
+const bytes32ContractsCache: Record<string, Contract> = {}
+
+const RETRY_OPTIONS = { n: 3, minWait: 100, maxWait: 3_000 }
+
+/**
+ * Hook that returns a callback which fetches data for a single token from the chain
+ *
+ * Alternative to hooks/Token/useToken where token address does not need to be known
+ * at hook phase nor uses multicall
+ */
+export function useTokenLazy() {
+  const { library, account, chainId } = useActiveWeb3React()
+  const addUserToken = useAddUserToken()
+
+  return useCallback(
+    async (address: string): Promise<Token | null> => {
+      console.debug(`[useTokenLazy::callback] callback called`, address, account, chainId)
+
+      if (!account || !chainId || !address) {
+        return null
+      }
+
+      console.debug(`[useTokenLazy::callback] input is valid`, address, account, chainId)
+
+      contractsCache[address] =
+        contractsCache[address] || getTokenContract(address, undefined, library, account, chainId)
+      bytes32ContractsCache[address] =
+        bytes32ContractsCache[address] || getBytes32TokenContract(address, undefined, library, account, chainId)
+
+      const contract = contractsCache[address]
+      const bytes32Contract = bytes32ContractsCache[address]
+
+      // Fetch each property individually
+      const { promise: decimalsPromise } = retry(contract.decimals, RETRY_OPTIONS)
+      const { promise: namePromise, cancel: cancelNamePromise } = retry(contract.name, RETRY_OPTIONS)
+      const { promise: symbolPromise, cancel: cancelSymbolPromise } = retry(contract.symbol, RETRY_OPTIONS)
+
+      let decimals: number
+
+      try {
+        // If no decimals, stop here
+        decimals = await decimalsPromise
+      } catch (e) {
+        console.debug(`[useTokenLazy::callback] no decimals, stopping`, address, account, chainId, e)
+
+        cancelNamePromise()
+        cancelSymbolPromise()
+        return null
+      }
+
+      const [nameSettled, symbolSettled] = await Promise.allSettled([namePromise, symbolPromise])
+
+      let name, symbol
+
+      // If no name or symbol, try the bytes32 contract
+      if (
```

Comment:
The whole useEffect is extremely complex

I did #1857 #1858 to experiment a bit and try to refactor. Feel free to improve it here, or in other PR. In general my take aways are:
- Complex logic should be broken down into smaller pieces
- retry logic makes things a bit complex because of its signature viral nature (works very bad with chaining of retry logic)
- probably this libraries like SWR would help with these things
- ideally we would delay the creation of the contract for byte32 so we only create it if we need it